src/librustc_data_structures/sorted_map.rs

```diff
@@ -247,38 +260,77 @@ impl<K: Ord, V> SortedMap<K, V> {
 
         (start, end)
     }
+
+    #[inline]
+    pub fn contains_key<Q>(&self, key: &Q) -> bool
+        where K: Borrow<Q>,
+              Q: Ord + ?Sized
+    {
+        self.get(key).is_some()
+    }
+}
+
+pub struct Iter<'a, K: Ord + 'a, V: 'a>(::std::slice::Iter<'a, (K, V)>);
+
+impl<'a, K: Ord + 'a, V: 'a> Iterator for Iter<'a, K, V> {
+    type Item = (&'a K, &'a V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((k, v)) = self.0.next() {
+            Some((&k, &v))
+        } else {
+            None
+        }
+    }
```

Comment:
Can we add size_hint here? It's important for every iterator implementation, in particular here when we have an exact one to use from the inner iterator.

One might consider, unless there is a pressing need, to keep using the slice iterator and not wrap it; then you don't have to worry about wrapping other iterator methods down the line, even if it would only be a few for the slice iterator.

Reply:
`size_hint` is a good idea, sure.

I wrapped the slice iterator in order to have its `Item` compatible with the one in `BTreeMap`; for slice the `Item` is `&'a T`, which for `SortedMap` is `&'a (K, V)`, while for `BTreeMap` the `Item` is `(&'a K, &'a V)`. Is there a better way to keep them compatible?

Reply:
IntoIterator you mean.

I can't answer this question since I'm not really driving anything behind the rustc datastructures, so I don't know how important it is to make them full fledged APIs.

Reply:
I don't see how this can work without using `existential type` (which may not work).

If you can't/don't want to get that to work, just manually implement `size_hint` and `try_fold` (to match `Map`) and call it a day.